game/src/ltn/algorithms.rs

```diff
@@ -0,0 +1,146 @@
+            interior,
+            perimeter,
+            borders,
+
+            modal_filters: BTreeSet::new(),
+            rat_runs: Vec::new(),
+        };
+        n.rat_runs = n.find_rat_runs(map);
+        n
+    }
+
+    pub fn toggle_modal_filter(&mut self, map: &Map, r: RoadID) {
+        if self.modal_filters.contains(&r) {
+            self.modal_filters.remove(&r);
+        } else {
+            self.modal_filters.insert(r);
+        }
+        self.rat_runs = self.find_rat_runs(map);
+    }
+
+    pub fn is_interior_road(r: RoadID, map: &Map) -> bool {
+        let road = map.get_r(r);
+        road.get_rank() == RoadRank::Local
+            && road
+                .lanes
+                .iter()
+                .any(|l| PathConstraints::Car.can_use(l, map))
+    }
+
+    // Just returns a sampling of rat runs, not necessarily all of them
+    fn find_rat_runs(&self, map: &Map) -> Vec<RatRun> {
+        // Just flood from each border and see if we can reach another border.
+        //
+        // We might be able to do this in one pass, seeding the queue with all borders. But I think
+        // the "visited" bit would get tangled up between different possibilities...
+        let mut runs: Vec<RatRun> = self
+            .borders
+            .iter()
+            .flat_map(|i| self.rat_run_from(map, *i))
+            .collect();
+        runs.sort_by(|a, b| a.length_ratio.partial_cmp(&b.length_ratio).unwrap());
+        runs
+    }
+
+    fn rat_run_from(&self, map: &Map, start: IntersectionID) -> Option<RatRun> {
+        // We don't need a priority queue
+        let mut back_refs = HashMap::new();
+        let mut queue = vec![start];
+
+        while !queue.is_empty() {
+            let current = queue.pop().unwrap();
+            if current != start && self.borders.contains(&current) {
+                // Found one!
+                let mut at = current;
+                let mut path = vec![at];
+                while let Some(prev) = back_refs.remove(&at) {
+                    path.push(prev);
+                    at = prev;
+                }
+                path.push(start);
+                path.reverse();
+                return Some(RatRun::new(path, map));
+            }
+
+            for r in &map.get_i(current).roads {
+                let next = map.get_r(*r).other_endpt(current);
+                if !self.interior.contains(r)
+                    || self.modal_filters.contains(r)
+                    || back_refs.contains_key(&next)
+                {
+                    continue;
+                }
+                back_refs.insert(next, current);
+                queue.push(next);
+            }
+        }
+
+        None
+    }
+}
+
+impl RatRun {
+    fn new(path: Vec<IntersectionID>, map: &Map) -> RatRun {
+        let mut run = RatRun {
+            path,
+            length_ratio: 1.0,
+        };
+        if let Some((roads, _)) = map.simple_path_btwn(run.path[0], *run.path.last().unwrap()) {
```

Comment:
Since we find so many possible rat runs, we want to rank them by how likely / useful to drivers they are. So we just measure the length of the path of the rat-run, then do regular pathfinding on the entire map, and compare the ratio. Still some issues with this heuristic, but it's an easy start.